### PR TITLE
Fix/binary crossentropy fn smoothing

### PR DESCRIPTION
**Pull request by kaivalya-cyber**

### CODE REVIEW

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @kaivalya-cyber,

Thank you for your contribution!

I reviewed the PR and the proposed changes to `categorical_crossentropy` label smoothing.

There is a critical issue in the current implementation of `categorical_crossentropy` when `axis` is not `-1` (default) and `label_smoothing > 0`.

Currently, `num_classes` is hardcoded to be derived from the last dimension of `y_true`:
```python
num_classes = math_ops.cast(array_ops.shape(y_true)[-1], y_pred.dtype)
```

If a user specifies a different `axis` (e.g., `axis=0` for channels-first data), `y_true` might have shape `[num_classes, batch_size]`. In this case, `array_ops.shape(y_true)[-1]` will return the batch size instead of the number of classes. This leads to:
1. Incorrect label smoothing value (`label_smoothing / num_classes`).
2. Targets that do not sum to 1, causing value distortions in loss computation.

### Verification Test

Here is a test case verifying this bug. It compares the loss computed by `categorical_crossentropy` with `axis=0` against manual computations with correct and buggy smoothing.

```python
import tensorflow as tf
import numpy as np

class LossesTest(tf.test.TestCase):

  def test_categorical_crossentropy_axis(self):
    # Shape [num_classes, batch_size] = [3, 2]
    y_true = np.array([[1.0, 0.0],
                       [0.0, 1.0],
                       [0.0, 0.0]], dtype=np.float32)
    y_pred = np.array([[0.8, 0.1],
                       [0.1, 0.8],
                       [0.1, 0.1]], dtype=np.float32)
    
    label_smoothing = 0.1
    
    # Expected smoothed y_true if num_classes=3:
    expected_y_true_axis_0 = y_true * 0.9 + 0.1 / 3.0
    
    # If code uses shape[-1] = 2:
    buggy_y_true = y_true * 0.9 + 0.05
    
    # Call with axis=0
    loss_axis_0 = tf.keras.losses.categorical_crossentropy(
        y_true, y_pred, axis=0, label_smoothing=label_smoothing
    )
    
    # Manual computation with buggy smoothing (div by 2)
    buggy_loss = tf.keras.losses.categorical_crossentropy(
        buggy_y_true, y_pred, axis=0, label_smoothing=0.0
    )
    
    # Manual computation with correct smoothing (div by 3)
    correct_loss = tf.keras.losses.categorical_crossentropy(
        expected_y_true_axis_0, y_pred, axis=0, label_smoothing=0.0
    )
    
    # Verify that current implementation yields the buggy loss
    self.assertAllClose(loss_axis_0, buggy_loss)
    # Verify that it does NOT yield the correct loss
    self.assertNotAllClose(loss_axis_0, correct_loss)

if __name__ == '__main__':
  tf.test.main()
```

### Proposed Fix

Change the `num_classes` extraction to use the specified `axis`:
```python
num_classes = math_ops.cast(array_ops.shape(y_true)[axis], y_pred.dtype)
```

This ensures that the correct dimension is used for determining the number of classes, regardless of whether `axis` is `-1` or any other valid dimension.

Could you please update the PR to include this fix and regression tests for non-default `axis` with label smoothing?

Thanks!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
## Potential Improvements / Action Items

We noticed that the new test file `losses_test.py` was added without corresponding updates to the build system in the PR.

If your repository setup (e.g., Bazel/BUILD files) requires explicit test targets, please ensure you add a `py_test` target for `losses_test.py`.

Example target configuration (if using Bazel):

```python
py_test(
    name = "losses_test",
    srcs = ["losses_test.py"],
    python_version = "PY3",
    srcs_version = "PY3",
    deps = [
        ":losses",
        # Add other necessary dependencies here
    ],
)
```

Also, ensure that `losses` target has all necessary dependencies (like `tf_utils` or equivalent if they are not picked up automatically).

**Comment by kaivalya-cyber** [comment]
@dmiltr3 thanks for the detailed review! Addressed both issues:
 
**1. Critical fix: `num_classes` now uses `axis` instead of `[-1]`**
 
In `categorical_crossentropy`, `_smooth_labels` was using `array_ops.shape(y_true)[-1]`
to determine `num_classes`, which breaks when `axis != -1` (e.g., channels-first data
with `axis=0` would use batch_size instead of num_classes). Fixed to use
`array_ops.shape(y_true)[axis]`.
 
**2. Robust label_smoothing validation for `binary_crossentropy`**
 
Added try/except guard around the `0 <= label_smoothing <= 1` check to catch
`TypeError` (None, strings) and `ValueError` (non-scalar types) and raise clear
error messages instead of cryptic ones.
 
**3. Regression test & BUILD target**
 
- Added `CategoricalCrossentropyAxisTest.testLabelSmoothingWithAxis` verifying
  correct label smoothing with `axis=0` (matches correct smoothing, not buggy)
- Added `py_test` target for `losses_test.py` in `tensorflow/python/keras/BUILD`

**Review by dmiltr3 — APPROVED** [review]

**Review by dmiltr3 — APPROVED** [review]
## Note

It appears you are modifying code in `tensorflow/python/keras`. This path is considered stale/legacy. Modern Keras development typically happens in a standalone repository (e.g., `keras-team/keras`).

Please verify if your changes should be directed to the standalone Keras repository instead, as changes directly to `tensorflow/python/keras` might not be integrated into modern Keras versions and could be overwritten or ignored in future updates.
